File: src/resolver/twitter.rs
use std::sync::Arc;

use anyhow::{Context, Result, anyhow};
use reqwest::{
    Client, Response, Url,
    header::{HeaderMap, HeaderValue},
};
use serde::Deserialize;
use tokio::sync::RwLock;

use crate::resolver::{Resolver, ResolverResult};

pub struct TwitterResolver {
    client: Arc<Client>,
    guest_token: RwLock<Option<String>>,
    common_headers: HeaderMap,
}

impl TwitterResolver {
    pub fn new(client: Arc<Client>) -> Self {
        let mut common_headers = HeaderMap::new();
        common_headers.insert(
            "User-Agent",
            HeaderValue::from_static("Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:122.0) Gecko/20100101 Firefox/122.0")
        );
        common_headers
            .insert("Authorization", HeaderValue::from_static("Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA"));
        common_headers.append("x-twitter-client-language", HeaderValue::from_static("en"));
        common_headers.append("x-twitter-active-user", HeaderValue::from_static("yes"));
        common_headers.append("accept-language", HeaderValue::from_static("en"));
        return Self {
            client,
            guest_token: RwLock::new(None),
            common_headers,
        };
    }

    async fn get_guest_token(&self) -> Result<String> {
        let res = self
            .client
            .post("https://api.x.com/1.1/guest/activate.json")
            .headers(self.common_headers.clone())
            .send()
            .await
            .context("Twitter:get_guest_token: cannot make request")?
            .bytes()
            .await
            .context("Twitter:get_guest_token: cannot read body")?;
        let json: TwitterGuestTokenResponse =
            serde_json::from_slice(&res).context("Twitter:get_guest_token: cannot parse json")?;
        return Ok(json.guest_token.into());
    }

    async fn do_request(&self, url: Url) -> Result<Response> {
        let guest_token = self
            .guest_token
            .read()
            .await
            .clone()
            .context("Twitter:do_request: guest_token is empty")?;
        return self
            .client
            .get(url)
            .headers(self.common_headers.clone())
            .header("x-guest-token", guest_token)
            .send()
            .await
            .with_context(|| format!("Twitter:do_request: cannot make request"));
    }
}

impl Resolver for TwitterResolver {
    async fn resolve(&self, source_url: &Url) -> Result<ResolverResult> {
        if self.guest_token.read().await.is_none() {
            let token = self
                .get_guest_token()
                .await
                .context("twitter:resolve: cannot get guest_token")?;
            let mut guard = self.guest_token.write().await;
            *guard = Some(token);
        }
        let tweet_id = source_url
            .path_segments()
            .context("Twitter:resolve: cannot get path segments")?
            .last()
            .context("Twitter:resolve: cannot get tweet id")?;
        let url = Url::parse_with_params(
            "https://x.com/i/api/graphql/2ICDjqPd81tulZcYrtpTuQ/TweetResultByRestId",
            [(
                "variables",
                format!(
                    r#"{{"tweetId":"{tweet_id}","includePromotedContent":false,"withVoice":false,"withCommunity":false}}"#
                ),
            )],
        )?;
        let mut res = self
            .do_request(url.clone())
            .await
            .context("Twitter:resolve: cannot do request")?;
        let status_code = res.status();
        if status_code == 403 || status_code == 404 || status_code == 429 {
            let mut guard = self.guest_token.write().await;
            *guard = Some(
                self.get_guest_token()
                    .await
                    .context("Twitter:resolve: cannot get guest_token")?,
            );
            res = self
                .do_request(url)
                .await
                .context("Twitter:resolve: cannot do request")?;
        }
        if status_code != 200 {
            return Err(anyhow!(
                "Twitter:resolve: bad response status code: {status_code}"
            ));
        }
        let body = res
            .bytes()
            .await
            .context("Twitter:resolve: cannot read body")?;
        let json: TwitterResponse =
            serde_json::from_slice(&body).context("Twitter:resolve: cannot parse json")?;
        let media = json
            .data
            .tweet_result
            .result
            .legacy
            .entities
            .media
            .get(0)
            .context("Twitter:resolve: media is empty")?;
        let url = media
            .video_info
            .variants
            .iter()
            .filter(|v| v.content_type == "video/mp4")
            .max_by_key(|v| v.bitrate)
            .map(|v| v.url.clone())
            .context("Twitter:resolve: cannot find url")?;
        return Ok(ResolverResult {
            url,
            width: media.original_info.width,
            height: media.original_info.height,
            referer: Some("https://x.com".to_string()),
        });
    }
}

#[derive(Deserialize)]
struct TwitterGuestTokenResponse {
    guest_token: String,
}

#[derive(Deserialize)]
struct TwitterResponse {
    data: TwitterData,
}

#[derive(Deserialize)]
struct TwitterData {
    #[serde(rename = "tweetResult")]
    tweet_result: TwitterTweetResult,
}

#[derive(Deserialize)]
struct TwitterTweetResult {
    result: TwitterResult,
}

#[derive(Deserialize)]
struct TwitterResult {
    legacy: TwitterLegacy,
}

#[derive(Deserialize)]
struct TwitterLegacy {
    entities: TwitterEntities,
}

#[derive(Deserialize)]
struct TwitterEntities {
    media: Vec<TwitterMedia>,
}

#[derive(Deserialize)]
struct TwitterMedia {
    original_info: TwitterOriginalInfo,
    video_info: TwitterVideoInfo,
}

#[derive(Deserialize)]
struct TwitterOriginalInfo {
    width: u32,
    height: u32,
}

#[derive(Deserialize)]
struct TwitterVideoInfo {
    variants: Vec<TwitterVariant>,
}

#[derive(Debug, Deserialize)]
struct TwitterVariant {
    content_type: String,
    bitrate: Option<u64>,
    url: String,
}

File: src/parsers/twitter.ts
import { HttpClient, HttpClientResponse } from "@effect/platform";
import {
  Array as A,
  Context,
  Effect,
  Layer,
  Option,
  Order,
  Schedule,
  Schema,
} from "effect";

class TwitterError extends Schema.TaggedError<TwitterError>()("TwitterError", {
  message: Schema.String,
}) { }

const GuestTokenResponse = Schema.Struct({
  guest_token: Schema.String,
});

const VideoInfoVariant = Schema.Struct({
  bitrate: Schema.OptionFromUndefinedOr(Schema.Int),
  content_type: Schema.String,
  url: Schema.String,
});
type VideoInfoVariant = typeof VideoInfoVariant.Type;
type VideoInfoVariantWithBitrate = VideoInfoVariant & {
  bitrate: Option.Some<number>;
};

const TwitterResponse = Schema.Struct({
  data: Schema.Struct({
    tweetResult: Schema.Struct({
      result: Schema.Struct({
        legacy: Schema.Struct({
          entities: Schema.Struct({
            media: Schema.Array(
              Schema.Struct({
                type: Schema.String,
                video_info: Schema.Struct({
                  variants: Schema.Array(VideoInfoVariant),
                }),
              }),
            ),
          }),
        }),
      }),
    }),
  }),
});

const isMp4Variant = (v: VideoInfoVariant): v is VideoInfoVariantWithBitrate =>
  v.content_type === "video/mp4" && Option.isSome(v.bitrate);

const VideoInfoVariantByBitrateOrder = Order.mapInput(
  Order.number,
  (v: VideoInfoVariantWithBitrate) => v.bitrate.value,
);

const commonHeaders = {
  "User-Agent":
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:122.0) Gecko/20100101 Firefox/122.0",
  Authorization:
    "Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA",
  "x-twitter-client-language": "en",
  "x-twitter-active-user": "yes",
  "accept-language": "en",
};

const guestTokenRetryPolicy = Schedule.exponential("50 millis").pipe(
  Schedule.jittered,
  Schedule.intersect(Schedule.recurs(4)),
);
const twitterRetryPolicy = Schedule.exponential("50 millis").pipe(
  Schedule.jittered,
  Schedule.intersect(Schedule.recurs(4)),
);

class TwitterGuestTokenCached extends Context.Tag("TwitterGuestTokenCached")<
  TwitterGuestTokenCached,
  {
    readonly get: () => Effect.Effect<string, TwitterError>;
  }
>() { }

export const TwitterGuestTokenCachedLive = Layer.effect(
  TwitterGuestTokenCached,
  Effect.gen(function*() {
    const cached = yield* Effect.cachedWithTTL(
      Effect.gen(function*() {
        const client = yield* HttpClient.HttpClient;
        const res = yield* client
          .post("https://api.x.com/1.1/guest/activate.json", {
            headers: commonHeaders,
          })
          .pipe(Effect.retry(guestTokenRetryPolicy));
        const body =
          yield* HttpClientResponse.schemaBodyJson(GuestTokenResponse)(res);
        return body.guest_token;
      }).pipe(
        Effect.tapErrorCause(Effect.logError),
        Effect.catchAll(
          (_) => new TwitterError({ message: `Cannot get guest token` }),
        ),
      ),
      "5 minutes",
    );
    return { get: () => cached };
  }),
);

export const getVideoUrl = (sourceUrl: URL) =>
  Effect.gen(function*() {
    const tokenCached = yield* TwitterGuestTokenCached;
    const token = yield* tokenCached.get();
    const tweetId = yield* A.last(sourceUrl.pathname.split("/")).pipe(
      Effect.mapError(
        (_) => new TwitterError({ message: "Cannot find tweet id" }),
      ),
    );
    const url = new URL(
      "https://x.com/i/api/graphql/2ICDjqPd81tulZcYrtpTuQ/TweetResultByRestId",
    );
    url.searchParams.set(
      "variables",
      JSON.stringify({
        tweetId,
        includePromotedContent: false,
        withVoice: false,
        withCommunity: false,
      }),
    );
    const client = yield* HttpClient.HttpClient;
    const res = yield* client
      .get(url, {
        headers: {
          ...commonHeaders,
          "x-guest-token": token,
        },
      })
      .pipe(
        Effect.retry(twitterRetryPolicy),
        Effect.tapErrorCause(Effect.logError),
        Effect.catchTags({
          RequestError: () =>
            new TwitterError({ message: "Cannot make request" }),
          ResponseError: () =>
            new TwitterError({ message: "Twitter response error" }),
        }),
      );
    const data = yield* HttpClientResponse.schemaBodyJson(TwitterResponse)(
      res,
    ).pipe(
      Effect.tapErrorCause(Effect.logError),
      Effect.mapError(
        (_) => new TwitterError({ message: "Cannot parse response" }),
      ),
    );
    const video = yield* A.findFirst(
      data.data.tweetResult.result.legacy.entities.media,
      (v) => v.type === "video",
    ).pipe(
      Effect.mapError(
        (_) => new TwitterError({ message: "Cannot find media of type video" }),
      ),
      Effect.map((v) => A.filter(v.video_info.variants, isMp4Variant)),
      Effect.flatMap((v) =>
        A.match(v, {
          onEmpty: () =>
            Effect.fail(new TwitterError({ message: "Cannot find video" })),
          onNonEmpty: Effect.succeed,
        }),
      ),
      Effect.map((v) => A.max(v, VideoInfoVariantByBitrateOrder)),
    );
    return video.url;
  });
